src/Microsoft.Azure.WebJobs.Host/Bindings/Path/BindingTemplateToken.cs

```diff
@@ -1,32 +1,293 @@
+            {
+                _expressionParts = expressionParts;
+                _builtin = builtin;
+            }
+
+            public override string AsLiteral => null;
+            
+            public override string ParameterName
+            {
+                get
+                {
+                    return this._expressionParts[0];
+                }
+            }
+
+            protected override void BuildCapturePattern(StringBuilder builder)
+            {
+                if (this._expressionParts.Length != 1)
+                {
+                    throw new InvalidOperationException($"Capture expressions can't include dot operators");
+                }
+
+                var value = this._expressionParts[0];
+                builder.Append(String.Format(CultureInfo.InvariantCulture, "(?<{0}>.*)", value));
+            }
+
+            public override string Evaluate(IReadOnlyDictionary<string, object> bindingData)
+            {
+                object current;
+
+                if (bindingData.TryGetValue(this._expressionParts[0], out current))
+                {
+                    for (int i = 1; i < _expressionParts.Length; i++)
+                    {
+                        var propName = _expressionParts[i];
+
+                        try
+                        {
+                            current = GetProperty(current, propName);
+                        }
+                        catch (Exception e)
+                        {
+                            throw new InvalidOperationException($"Error while accessing '{propName}': {e.Message}");
+                        }
+                    }
+                }
+                else
+                {
+                    // Not found in binding data.  Is this a builtin?
+                    if (this._builtin != null)
+                    {
+                        current = _builtin.Resolve(this.ParameterName);
+                    }
+
+                    if (current == null)
+                    {
+                        throw new InvalidOperationException(String.Format(CultureInfo.CurrentCulture, "No value for named parameter '{0}'.", this.ParameterName));
+                    }
+                }
+
+                var strValue = BindingDataPathHelper.ConvertParameterValueToString(current);
+                return strValue;
+            }
+
+            // Support:
+            //   IDictionary<string, T>;
+            //   IReadOnlyDictionary<string, T>
+            // There's no common base interface here, and T can be { string, object, poco }, 
+            // so we just directly reflect over 'bool TryGetValue(string, out T)' and invoke that. 
+            private static bool TryGetValue(object o, string member, out object result)
+            {
+                Type scope = o.GetType();
+
+                var method = scope.GetMethod("TryGetValue", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
```

Comment:
Matt pointed it out before, that this is a lot of reflection to do for every single property access. Since the 95% case will be one of IDictionary<string, string>, IDictionary<string, object>, or a read only variety, I wonder if we shouldn't special case those?

Another idea might be to use the indexer:
```csharp
private static bool TryGetIndexedValueOrNull(object target, string name, out object value)
{
    value = null;

    var indexer = target.GetType().GetProperties()
        .Where(p => p.Name == "Item" && 
                p.GetIndexParameters().Count() == 1 && 
                p.GetIndexParameters()[0].ParameterType == typeof(string))
        .FirstOrDefault();
            
    if (indexer != null)
    {
        try
        {
            value = indexer.GetValue(target, new object[] { name });
            return true;
        }
        catch
        {
        }
    }

    return false;
}
```